### ISSUE TITLE
deletePayment - is it delete, or update?

### ISSUE THREAD

**Issue by judgej** [XeroAPI/Xero-OpenAPI#89]
I have been looking at the `deletePayment` operation ID, and some things about it are a little unclear to me. It seems to not quite make up its mind whether it is an update or a delete, and that results in some inconsistencies in code generated from the API description. I think it needs to go one way or the other.

The `deletePayment` operation is a `POST` that takes a payment ID as a path parameter. Putting aside the fact that it's not a `DELETE` method, that should be all that is needed to instruct a deletion.

However, the operation then requires a `#/components/schemas/Payments` schema as a payload. That is the same schema needed to create a payment in the first place, so it seems that this delete is actually an update. Except that only the status can be changed (from what I can see when experimenting), and the status can only be changed to `DELETED`.

This is given as an example in the API description:

```json
{
    "Payments": [
        {
            "Status": "DELETED"
        }
    ]
}
```

I'm not even sure if this example is correct, since it shows an *array* of payments, none of which are individually identified. This below works in practice though, not the example:

```json
{
   "Status": "DELETED"
}
```

So:

* If this *is* a `deletePayment` operation, that only deletes payments, then the payload should perhaps be a fixed (constant) structure and not defined as a whole Payment schema. That would help in automated code generation immensely. However, so far as OpenAPI is concerned, you have "taken up" this method and endpoint for this single operation, and in theory - for a valid OpenAPI description - could not use it again if you wanted another operation there - i.e. updating, undeleting.
* If this is actually an update payment operation, that happens to only allow the status to be updated *for now*, then maybe it should be renamed to `updatePayment`. That would help to understand how this endpoint works.

I've not tried to see if the operation can be used to undelete a deleted payment, and if it can, it would open up further questions.

I'm happy to provide a PR, but need to know what the original intention was here.

**Comment by judgej** [comment]
Looking again at my generated code, it does pick up the payment as an array of Payments, which is wrong - it's a single payment. The example *is* consistent with the description, but both are wrong IMO.

**Comment by SidneyAllen** [comment]
Hi @judgej,

We are currently generating our Java SDK off the OAS and I've just tested the deletePayment operation with the payload as an array of payments and it functions properly.

We could remove this operation and let developers figure out that they need to perform an updatePayment call with the status=DELETED, but deleting payments is a commonly used operations, and we felt it would help developer by adding it the OAS as a separate operation.

Can you share what language you are working with and they exact problem this operation is introducing into your generated code?

**Comment by judgej** [comment]
I think there are a number of issues that are kind of contradictory. It hints at going one direction, but heads off in a number of other directions, so my main aim is to seek clarification, and maybe help improve the OpenAPI description. I *can* get it to work with the generated code, but it feels very fudgy.

I try to break it down:

* The path is `/Payments/{PaymentID}` so I would expect this endpoint to handle just one payment, not multiple payments.
* The payload, according to the OpenAPI description takes an array of payments (more accurately an object containing an array of payments). The implication is that this endpoint operates on *multiple* payments and not a single payment.
* The method is `POST` and not `DELETE`. I can understand this is just a legacy thing that won't be changing any time soon. It does make it more difficult to introduce an `updatePayment` operation in the future, since this method+path is now already taken up by the `deletePayment` operation. You may have no plans to do that, but as before, it is just clarification I am looking for when things look a little contradictory.
* The inner details of the payload for the `deletePayment` operation is the `Payment` schema. However, only the `status` can be set, and *must* be set (in the docs and in practice in my experience). It seems that isn't the `Payment` scheme, but a different scheme - `deletePayment` schema maybe, containing just the mandatory `status` field? Going a step further, if the payment *must* contain a `payments` element with an array with a single `deletePayment` schema in, then why not make that outer wrapper a part of that?

So I understand you would like to make it easier for developers, and that makes sense and is great, but in doing so there is still a crazy dance of schemas to follow through what looks like an `update` operation in order to perform a `delete` operation.

**Comment by judgej** [comment]
I'm using the generated PHP code here to delete a payment. The resulting data structures are reallyu not intuitive, and as a developer need some research:

https://github.com/consilience/xero-api-sdk/tree/master/apis/accounting/src/Api

```php
use Consilience\Xero\AccountingSdk\Api;
use Consilience\Xero\AccountingSdk\Model\Payment;
use Consilience\Xero\AccountingSdk\Model\Payments;

$config = ...
$api = AccountingApi($config);
$dummyUpdatePayment = new Payment([
    'status' = Payment::STATUS_DELETED
]);
$dummyUpdatePayments = new Payments([
    'payments' => [$dummyUpdatePayment]
]);
$api = deletePayment($myPaymentId, $dummyUpdatePayments);
```

Ideally I would just do this, and it is possible to structure the OpenAPI description so this is supported out of the box:

    $api = deletePayment($myPaymentId);

**Comment by SidneyAllen** [comment]
Hi @judgej 

UpdatePayment is not on our roadmap at this point.   As always, thanks for sharing your POV.  Xero is working with these OpenAPI specs to generate our own official libraries, so any changes need to be validated against our current client SDKs and those in development.

The current deletePayment functions properly, so this isn't an urgent issue.  With that said, we love hearing how developers are using our OAS in the wild and how we can further polish it to make it better.   If you want to submit a PR to achieve the following outcome, we'd be interested in seeing it.

$deletedPayment = $api->deletePayment($myPaymentId);

I'll tag this as an enhancement and keep the issue open.

**Comment by SidneyAllen** [comment]
@judgej 

I've circled back on this issue - this isn't going to be solved at the API level, so I've modified the specification to make it a little better (not perfect)

and have added a "PaymentDelete" object with a single property "Status" with a default value of "DELETED".

I've tested in our Java SDK and now you only need to create a new instance of the PaymentDelete object to pass as an arg. to the deletePayment operation.

PaymentDelete paymentDelete = new PaymentDelete();
Payments deletedPayment = accountingApi.deletePayment(accessToken, xeroTenantId, paymentID, paymentDelete);
             
This work is in a branch and will be merged in release 2.0.5 of our OAS.
